### PR TITLE
Implemented various versions of trash, as well as a trash factory. Re…

### PR DESCRIPTION
**Pull request by ibuenros**
This PR includes:
* Various new Trash implementations
** ProxiedTrash
** MockTrash: for simulate
** ImmediateDeletionTrash: for skip trash
** TestTrash: for an instance used for testing
* TrashFactory to automatically select correct trash implementation based on properties.
* Many trash tests.
* Whitelist predicate for Dataset cleaner.

### CODE REVIEW

**Comment by ibuenros** [comment]
@ydailinkedin can you review?

File: gobblin-data-management/src/main/java/gobblin/data/management/trash/ProxiedTrash.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  /**
+   * Get {@link gobblin.data.management.trash.Trash} instance for the specified user.
+   * @param user user for whom {@link gobblin.data.management.trash.Trash} should be generated.
+   * @return {@link gobblin.data.management.trash.Trash} as generated by proxied user.
+   * @throws IOException
+   */
+  protected Trash getUserTrash(final String user) throws IOException {
+    if(UserGroupInformation.getCurrentUser().getUserName().equals(user)) {
```
Comment (ydailinkedin) [inline gobblin-data-management/src/main/java/gobblin/data/management/trash/ProxiedTrash.java]: Have you verified if `UserGroupInformation.getCurrentUser().getUserName()` simply returns the `[user-name]` or  `[user-name]/[cluster-name]@[realm]`?

File: gobblin-data-management/src/main/java/gobblin/data/management/trash/Trash.java [hunk 2]
```diff
@@ -69,17 +74,27 @@ public boolean accept(Path path) {
...
+   * Create location of Trash directory. Parsed from props at key {@link #TRASH_LOCATION_KEY}, defaulting to
    * /home/directory/_GOBBLIN_TRASH.
    * @param fs {@link org.apache.hadoop.fs.FileSystem} where trash should be found.
    * @param props {@link java.util.Properties} containing trash configuration.
+   * @param user If the trash location contains the token $USER, the token will be replaced by the value of user.
    * @return {@link org.apache.hadoop.fs.Path} for trash directory.
    * @throws java.io.IOException
    */
-  public static Path getTrashLocation(FileSystem fs, Properties props) throws IOException {
+  protected Path createTrashLocation(FileSystem fs, Properties props, String user) throws IOException {
```
Comment (ydailinkedin) [inline gobblin-data-management/src/main/java/gobblin/data/management/trash/Trash.java]: Thanks for this part! I was gonna make a similar change.

File: gobblin-utility/src/main/java/gobblin/util/ProxiedFileSystemWrapper.java [hunk 3]
```diff
@@ -120,7 +120,7 @@ public Void run() throws IOException {
    * @return Token for proxyUserName if it exists.
    * @throws IOException
    */
-  private Optional<Token> getTokenFromSeqFile(String authPath, String proxyUserName) throws IOException {
+  public static Optional<Token> getTokenFromSeqFile(String authPath, String proxyUserName) throws IOException {
```
Comment (ydailinkedin) [inline gobblin-utility/src/main/java/gobblin/util/ProxiedFileSystemWrapper.java]: This class has been deprecated. You can use `ProxiedFileSystemUtils.getTokenFromSeqFile` instead.

**Comment by ibuenros** [comment]
@ydailinkedin Addressed comments, and fixed a bug where new methods in ProxiedFileSystem cache were not actually using cache.

**Pull request merged**